File: benchmark/synduce/list/sumhom.f
Inductive CList = cnil Unit | single Int | concat {CList, CList};
Inductive List = nil Unit | cons {Int, List};

cat = fix (\f: List->List->List. \x: List. \y: List.
  match x with
    cons {h, t} -> cons {h, f t y}
  | nil _ -> y
  end
);

repr = fix (
  \f: CList -> Compress List. \cl: CList.
  match cl with
    cnil _ -> nil unit
  | single h -> cons {h, nil unit}
  | concat {l, r} -> cat (f l) (f r)
  end
);

spec = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> + h (f t)
  end
);

@Start main = \cl: CList. spec (repr cl);

File: ml-benchmark/synduce/list/sumhom.ml
type clist = CNil of unit | Single of int | Concat of clist * clist
type lst = Nil of unit | Cons of int * lst

let rec cat x y =
  match x with
  | Cons (h, t) -> Cons (h, cat t y)
  | Nil _ -> y

let rec repr cl =
  match cl with
  | CNil _ -> Nil ()
  | Single h -> Cons (h, Nil ())
  | Concat (l, r) -> cat (repr l) (repr r)

let rec spec xs =
  match xs with
  | Nil _ -> 0
  | Cons (h, t) -> h + spec t

let program cl = spec (repr cl)